artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPConnectionContext.java

```diff
@@ -389,13 +408,15 @@ public void onRemoteOpen(Connection connection) throws Exception {
       initialise();
 
       /*
-      * This can be null which is in effect an empty map, also we really don't need to check this for in bound connections
-      * but its here in case we add support for outbound connections.
-      * */
+       * This can be null which is in effect an empty map, also we really don't need to check this for in bound connections
+       * but its here in case we add support for outbound connections.
+       * */
       if (connection.getRemoteProperties() == null || !connection.getRemoteProperties().containsKey(CONNECTION_OPEN_FAILED)) {
          long nextKeepAliveTime = handler.tick(true);
-         if (nextKeepAliveTime != 0 && scheduledPool != null) {
-            scheduledPool.schedule(new ScheduleRunnable(), (nextKeepAliveTime - TimeUnit.NANOSECONDS.toMillis(System.nanoTime())), TimeUnit.MILLISECONDS);
+         synchronized (schedulingLock) {
```

Comment:
I have a very big concern here. Locks in connection stuff, artemis was built on the premise of being nonblocking. As time goes on more and more blocking locks are being added to core code paths. That then take months unravellling later. Can we please spend the time upfront now to find a solution to avoid adding sync locks.

Reply:
I am looking for a solution without sync locks but using AtomicBoolean and AtomicReference. Have you any suggestion?

Reply:
I created the PR #2727 to remove the synchronized blocks using an AtomicReference.